Skip Pro Features for Local Builds

### Is your feature request related to a problem? Please describe.

We'd like to build snapshots locally as close to production as possible (for things like installing packages in docker, etc) and we use a goreleaser target in our makefile to do so:

```

.PHONY: dist-backend
dist-backend:
	goreleaser release --config .goreleaser.private.yaml --clean --snapshot

.PHONY: dist-cli
dist-cli:
	goreleaser release --config .goreleaser.public.yaml --clean --snapshot
```

We just upgraded to goreleaser pro for the cloudsmiths repository hosting support.  However, now building locally we must manually comment out the cloudsmiths block otherwise we get this:

```
goreleaser release --config .goreleaser.private.yaml --clean --snapshot
  ⨯ release failed after 0s
    error=
    │ yaml: unmarshal errors:
    │   line 61: field cloudsmiths not found in type config.Project
make: *** [dist-backend] Error 1
```

### Describe the solution you'd like

We'd like an option or flag to skip the pro build features locally.  We really only want to upload to cloudsmiths from the CI jobs, and not from our local machines, so we'd like that step to be skipped entirely.

We don't want to have to distribute the license key for goreleaser pro to all our engineering team.

### Describe alternatives you've considered

We considered using multiple goreleaser files, but we already have 2 environment files targeting different packages but doing almost the same thing, so we'd like to avoid adding more for maintenance reasons.

### Search

- [x] I did search for other open and closed issues before opening this

### Supporter

- [x] I am a [sponsor](https://github.com/sponsors/caarlos0/) or a [Pro](https://goreleaser.com/pro) customer

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### Additional context

_No response_

--- a/.goreleaser-nightly.yaml
+++ b/.goreleaser-nightly.yaml
@@ -1,6 +1,7 @@
 # yaml-language-server: $schema=https://goreleaser.com/static/schema-pro.json
 # vim: set ts=2 sw=2 tw=0 fo=jcroql
 version: 2
+pro: true
 
 nightly:
   publish_release: true

--- a/cmd/build.go
+++ b/cmd/build.go
@@ -85,7 +85,7 @@ When using ` + "`--single-target`" + `, you use the ` + "`TARGET`, or GOOS`, `GO
 	cmd.Flags().StringArrayVar(&root.opts.ids, "id", nil, "Builds only the specified build ids")
 	_ = cmd.RegisterFlagCompletionFunc("id", func(_ *cobra.Command, _ []string, _ string) ([]string, cobra.ShellCompDirective) {
 		// TODO: improve this
-		cfg, err := loadConfig(root.opts.config)
+		cfg, err := loadConfig(!root.opts.snapshot, root.opts.config)
 		if err != nil {
 			return nil, cobra.ShellCompDirectiveNoFileComp
 		}
@@ -115,7 +115,7 @@ When using ` + "`--single-target`" + `, you use the ` + "`TARGET`, or GOOS`, `GO
 }
 
 func buildProject(options buildOpts) (*context.Context, error) {
-	cfg, err := loadConfig(options.config)
+	cfg, err := loadConfig(!options.snapshot, options.config)
 	if err != nil {
 		return nil, err
 	}

--- a/cmd/config.go
+++ b/cmd/config.go
@@ -10,10 +10,26 @@ import (
 	"github.com/goreleaser/goreleaser/v2/pkg/config"
 )
 
-func loadConfig(path string) (config.Project, error) {
+var proExplain = `Your configuration is for GoReleaser Pro.
+You are currently using GoReleaser OSS, so all the Pro-only features will be ignored.
+Use GoReleaser Pro to enable all the features.`
+
+func loadConfig(strict bool, path string) (config.Project, error) {
 	p, path, err := loadConfigCheck(path)
 	if err == nil {
 		log.WithField("path", path).Debug("using configuration")
+	}
+	if errors.Is(err, config.ErrProConfig) {
+		if strict {
+			return p, err
+		}
+		log.WithField("explanation", proExplain).
+			Warnf(
+				"%s %s",
+				logext.Warning("your configuration specifies"),
+				logext.Keyword("pro: true"),
+			)
+		return p, nil
 	}
 	return p, err
 }

--- a/cmd/config_test.go
+++ b/cmd/config_test.go
@@ -25,18 +25,49 @@ func TestConfigFlagNotSetButExists(t *testing.T) {
 				filepath.Join(folder, "goreleaser.yml"),
 				filepath.Join(folder, name),
 			))
-			proj, err := loadConfig("")
+			proj, err := loadConfig(true, "")
 			require.NoError(t, err)
 			require.NotEqual(t, config.Project{}, proj)
 		})
 	}
 }
 
+var proConfig = `version: 2
+pro: true
+some_possibly_pro_option: {}
+`
+
+func TestProConfigFile(t *testing.T) {
+	folder := setup(t)
+	require.NoError(t, os.WriteFile(
+		filepath.Join(folder, "goreleaser.yml"),
+		[]byte(proConfig),
+		0o644,
+	))
+	t.Run("strict", func(t *testing.T) {
+		proj, err := loadConfig(true, "goreleaser.yml")
+		require.Error(t, err)
+		require.Equal(t, config.Project{
+			Version: 2,
+			Pro:     true,
+		}, proj)
+	})
+
+	t.Run("relaxed", func(t *testing.T) {
+		proj, err := loadConfig(false, "goreleaser.yml")
+		require.NoError(t, err)
+		require.Equal(t, config.Project{
+			Version: 2,
+			Pro:     true,
+		}, proj)
+	})
+}
+
 func TestConfigFileDoesntExist(t *testing.T) {
 	folder := setup(t)
 	err := os.Remove(filepath.Join(folder, "goreleaser.yml"))
 	require.NoError(t, err)
-	proj, err := loadConfig("")
+	proj, err := loadConfig(true, "")
 	require.NoError(t, err)
 	require.Equal(t, config.Project{}, proj)
 }
@@ -45,7 +76,7 @@ func TestConfigFileFromStdin(t *testing.T) {
 	folder := setup(t)
 	err := os.Remove(filepath.Join(folder, "goreleaser.yml"))
 	require.NoError(t, err)
-	proj, err := loadConfig("-")
+	proj, err := loadConfig(true, "-")
 	require.NoError(t, err)
 	require.Equal(t, config.Project{}, proj)
 }

--- a/cmd/healthcheck.go
+++ b/cmd/healthcheck.go
@@ -38,7 +38,7 @@ func newHealthcheckCmd() *healthcheckCmd {
 				log.Log = log.New(io.Discard)
 			}
 
-			cfg, err := loadConfig(root.config)
+			cfg, err := loadConfig(true, root.config)
 			if err != nil {
 				return err
 			}

--- a/cmd/release.go
+++ b/cmd/release.go
@@ -103,7 +103,7 @@ func newReleaseCmd() *releaseCmd {
 }
 
 func releaseProject(options releaseOpts) (*context.Context, error) {
-	cfg, err := loadConfig(options.config)
+	cfg, err := loadConfig(options.snapshot, options.config)
 	if err != nil {
 		return nil, err
 	}

--- a/cmd/release.go
+++ b/cmd/release.go
@@ -103,7 +103,7 @@ func newReleaseCmd() *releaseCmd {
 }
 
 func releaseProject(options releaseOpts) (*context.Context, error) {
-	cfg, err := loadConfig(options.snapshot, options.config)
+	cfg, err := loadConfig(!options.snapshot, options.config)
 	if err != nil {
 		return nil, err
 	}

--- a/internal/yaml/yaml.go
+++ b/internal/yaml/yaml.go
@@ -13,7 +13,6 @@ import (
 func UnmarshalStrict(in []byte, out interface{}) error {
 	decoder := yaml.NewDecoder(bytes.NewReader(in))
 	decoder.KnownFields(true)
-
 	return handleErr(decoder.Decode(out))
 }
 
@@ -21,7 +20,6 @@ func UnmarshalStrict(in []byte, out interface{}) error {
 func Unmarshal(in []byte, out interface{}) error {
 	decoder := yaml.NewDecoder(bytes.NewReader(in))
 	decoder.KnownFields(false)
-
 	return handleErr(decoder.Decode(out))
 }
 
@@ -30,15 +28,11 @@ func Marshal(in interface{}) ([]byte, error) {
 	b := new(bytes.Buffer)
 	encoder := yaml.NewEncoder(b)
 	encoder.SetIndent(2) // default is 4
-
 	if err := encoder.Encode(in); err != nil {
 		encoder.Close()
-
 		return nil, err
 	}
-
 	encoder.Close()
-
 	return b.Bytes(), nil
 }
 
@@ -47,6 +41,5 @@ func handleErr(err error) error {
 	if err != nil && !errors.Is(err, io.EOF) {
 		return err
 	}
-
 	return nil
 }

--- a/pkg/config/config.go
+++ b/pkg/config/config.go
@@ -16,6 +16,7 @@ import (
 
 type Versioned struct {
 	Version int
+	Pro     bool
 }
 
 // Git configs.
@@ -1287,6 +1288,7 @@ type Source struct {
 // Project includes all project configuration.
 type Project struct {
 	Version         int              `yaml:"version,omitempty" json:"version,omitempty" jsonschema:"enum=2,default=2"`
+	Pro             bool             `yaml:"pro,omitempty" json:"pro,omitempty"`
 	ProjectName     string           `yaml:"project_name,omitempty" json:"project_name,omitempty"`
 	Env             []string         `yaml:"env,omitempty" json:"env,omitempty"`
 	Release         Release          `yaml:"release,omitempty" json:"release,omitempty"`

--- a/pkg/config/config_test.go
+++ b/pkg/config/config_test.go
@@ -121,3 +121,8 @@ func TestVersion(t *testing.T) {
 		require.NoError(t, err)
 	})
 }
+
+func TestPro(t *testing.T) {
+	_, err := LoadReader(strings.NewReader("version: 2\npro: true\nnope: true\n"))
+	require.ErrorIs(t, err, ErrProConfig)
+}

--- a/pkg/config/load.go
+++ b/pkg/config/load.go
@@ -1,6 +1,7 @@
 package config
 
 import (
+	"errors"
 	"fmt"
 	"io"
 	"os"
@@ -15,6 +16,10 @@ import (
 type VersionError struct {
 	current int
 }
+
+// ErrProConfig happens if the configuration failed to load strictly, but
+// there's a 'pro: true' field in it, so we just allow anything.
+var ErrProConfig = errors.New("you are using a GoReleaser Pro configuration file with GoReleaser OSS")
 
 func (e VersionError) Error() string {
 	return fmt.Sprintf(
@@ -54,6 +59,12 @@ func LoadReader(fd io.Reader) (config Project, err error) {
 	err = yaml.UnmarshalStrict(data, &config)
 	if err != nil && !validVersion {
 		return config, VersionError{versioned.Version}
+	}
+	if err != nil && versioned.Pro {
+		err2 := yaml.Unmarshal(data, &config)
+		if err2 == nil {
+			err = errors.Join(err, ErrProConfig)
+		}
 	}
 	return config, err
 }

--- a/www/docs/errors/version.md
+++ b/www/docs/errors/version.md
@@ -1,4 +1,6 @@
-# Unsupported configuration version
+# Version-related configuration errors
+
+## Unsupported configuration version
 
 This can show as an error or as an warning:
 
@@ -12,7 +14,7 @@ If you get it as a warning, your configuration file is valid in v2, but would
 benefit with the version header.
 You can remove the warning by adding this line to your configuration:
 
-```yaml
+```yaml title=".goreleaser.yml"
 version: 2
 ```
 
@@ -26,3 +28,22 @@ them.
 Also worth reading the
 [v2 announcement](../blog/posts/2024-06-04-goreleaser-v2.md), which contains an
 upgrade guide.
+
+## Using a Pro configuration file with GoReleaser OSS
+
+If you work with many people, you might not want to share the GoReleaser Pro key
+with everyone.
+
+Odds are, most of the people won't be doing much with it, maybe just testing the
+builds and things like that.
+
+If you want to allow GoReleaser OSS to read and use a GoReleaser Pro
+configuration, set `pro: true` in your configuration file:
+
+```yaml title=".goreleaser.yml"
+version: 2
+pro: true
+```
+
+Then, whenever `--snapshot` is set, GoReleaser will happily proceed without
+rejecting the configuration (it will warn about it, though).

--- a/www/docs/errors/version.md
+++ b/www/docs/errors/version.md
@@ -47,3 +47,8 @@ pro: true
 
 Then, whenever `--snapshot` is set, GoReleaser will happily proceed without
 rejecting the configuration (it will warn about it, though).
+
+!!! warning
+
+    When doing this, other YAML parsing errors might be ignored, such as fields
+    that don't actually exist.

